### PR TITLE
[improve](fold) support complex type for constant folding

### PR DESCRIPTION
**Pull request by zhangstar333**
## Proposed changes

wants support array/map/struct could constant fold, eg:

```
mysql [(none)]>desc select array_range(10,13);
+---------------------------------+
| Explain String(Nereids Planner) |
+---------------------------------+
| PLAN FRAGMENT 0                 |
|   OUTPUT EXPRS:                 |
|     [10, 11, 12][#0]            |
|   PARTITION: UNPARTITIONED      |
|                                 |
|   HAS_COLO_PLAN_NODE: false     |
|                                 |
|   VRESULT SINK                  |
|      MYSQL_PROTOCAL             |
|                                 |
|   0:VUNION(32)                  |
|      constant exprs:            |
|          [10, 11, 12]           |
+---------------------------------+
```

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhangstar333** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 1]
```diff
@@ -216,43 +217,11 @@ private Map<String, Expression> evalOnBE(Map<String, Map<String, TExpr>> paramMa
...
-                                        || primitiveType == PrimitiveType.DECIMAL256) {
-                                    type = DecimalV3Type.createDecimalV3TypeLooseCheck(
-                                            pScalarType.getPrecision(), pScalarType.getScale());
-                                } else {
-                                    type = DataType.fromCatalogType(ScalarType.createType(
-                                            PrimitiveType.fromThrift(tPrimitiveType)));
-                                }
-                                ret = Literal.of(e1.getValue().getContent()).castTo(type);
-                            }
+                            PTypeDesc pTypeDesc = e1.getValue().getPType();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:222]: why not use a tree to represent complex type? what's advantage of list? How to express if the elements in complex types support non-null in the future

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 2]
```diff
@@ -216,43 +217,11 @@ private Map<String, Expression> evalOnBE(Map<String, Map<String, TExpr>> paramMa
...
+                            DataType type = convertToNereidsType(pTypeDesc.getTypesList(), 0).key();
+                            ret = Literal.of(e1.getValue().getContent()).castTo(type);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:224]: i think, we'd better refactor fold constant on be. we should not rely on string as ser/de infra. we should use a better way to ser/de literal expression just like what we do in thrift.

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Review by 924060929 — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 3]
```diff
@@ -156,12 +189,23 @@ private static void collectConst(Expression expr, Map<String, Expression> constM
...
+                    || expr.getDataType().isIPv6Type()) {
                 return;
             }
-            if (skipSleepFunction(expr)) {
+            if (skipSleepFunction(expr) || (expr instanceof TableGeneratingFunction)) {
                 return;
             }
+            // Tokenize function want check the second child literal must be string type
+            // and properties format, it's a little special,
+            // maybe check in checkLegalityBeforeTypeCoercion function?
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:205]: agree with check in `checkLegalityBeforeTypeCoercion`

**Comment by 924060929** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 4]
```diff
@@ -291,4 +311,304 @@ private static Map<String, Expression> evalOnBE(Map<String, Map<String, TExpr>>
         }
         return resultMap;
     }
+
+    /**
+     * convert PValues which from BE to Expression of nereids
+     */
+    public static List<Literal> getResultExpression(DataType type, PValues resultContent) {
+        List<Literal> res = new ArrayList<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:319]: how about result is a NullLiteral with non-null type?

**Review by morrySnow — APPROVED** [review]

**Comment by WencongLiu** [comment]
@zhangstar333  Hi, R、recently, I have been developing some new functions and noticed that they are not being folded even when the constant folding feature is enabled. I would like to seek advice on how to modify them to achieve constant folding. Could you possibly add my WeChat, Liuwenclever, for further discussion? I would greatly appreciate your assistance. Thank you very much.

**Pull request merged**